### ISSUE TITLE
Marching Cubes Operation

### ISSUE THREAD

**Issue by drebain** [tensorflow/graphics#282]
It could be very useful to have a (potentially differentiable) marching cubes implementation for projects that work with implicit shape representations. Is this something that there would be interest for as a contribution?

**Comment by jackd** [comment]
I looked at this a while back and found it wasn't very useful in practice - at least for my research - and that my pure tf port was slower than wrapping existing implementations and using some dirty (albeit correct) gradient hacking. I haven't looked at it for quite a while so it'll need some cleaning up, but if you're starting from scratch feel free to take whatever you'd like from [here](https://github.com/jackd/tf_marching_cubes).

If nobody else gets around to it I'll probably put something together eventually, but I've got a fairly sizeable back-log of things to get through at the moment.

**Comment by julienvalentin** [comment]
Hi Daniel! Glad to hear from you!
This would indeed be a very interesting piece!

**Comment by julienvalentin** [comment]
Hi @drebain 

Are you still planning to contribute what you described above in this thread?

Best.

**Comment by drebain** [comment]
Yes. It seems like @jackd 's implementation above could be made into a tfg op, but I'm a bit worried about the performance. My experience with pure python tensorflow implementations that rely heavily on gather/scatter operations is that they seem to be much slower than an equivalent CUDA C++ implementation. I would be interested in doing a higher-performance version if you're accepting CUDA code.

Either way, I plan on starting this sometime in (now, NeurIPS deadline].

**Comment by drebain** [comment]
![marching_cubes_output](https://user-images.githubusercontent.com/5511462/82499216-26894380-9aa6-11ea-8b13-13d350742348.png)
**Update:** I have done a [first attempt at a C++/CUDA op](https://github.com/drebain/graphics/commit/194c5d61120e1262873f1abbfa0c879265e95609) and assuming I have done the benchmarking correctly, it seems like there is indeed a significant performance cost to the pure python implementation.

![marching_cubes_benchmark](https://user-images.githubusercontent.com/5511462/82499280-4456a880-9aa6-11ea-984f-109efd968a4a.png)
[[code](https://gist.github.com/drebain/7ea40f4f5c3d4a98ca5d8afc5593bff6)]

My implementation supports CPU and CUDA execution and implements the algorithm as described [here](http://paulbourke.net/geometry/polygonise/), producing a pure triangle soup instead of connecting the vertices within a cell or between cells. Ideally, there should be another op for merging duplicate vertices as this would also be useful for other tasks.

Aside from missing tests/docs/etc., there's also no build configuration as that doesn't seem to have been set up for C++ yet, though @ataiya tells me it's in progress.

**Comment by ataiya** [comment]
We can have both a pure python and an accelerated (C++ or CUDA version).
I don't think there is anything wrong with that ;)

The grid reso is sqrt(number of cells)?
Or are you counting the number of occupied cells
(out of curiosity)

**Comment by drebain** [comment]
It's the actual number of cells, this goes up to 128x128x128. It can go higher before it runs out of memory, but this was the most I had patience to wait for.

**Comment by drebain** [comment]
To be more specific it's the total number of cells in the grid. The occupied cells will be roughly half that, as I'm using Gaussian white noise with isolevel 0.0.
